core/transaction_pool/impl/transaction_pool_impl.cpp

```diff
@@ -0,0 +1,190 @@
+    waiting_queue_.erase(waiting_queue_.begin(), w_border);
+
+    for (auto &[hash, ready_tx] : ready_queue_) {
+      if (moderator_->banIfStale(number, ready_tx)) {
+        removed.push_back(ready_tx);
+        ready_queue_.erase(hash);
+      }
+    }
+
+    moderator_->updateBan();
+
+    return removed;
+  }
+
+  void TransactionPoolImpl::updateReady() {
+    auto is_ready = [this](auto &&tx) {
+      return std::all_of(
+          tx.requires.begin(), tx.requires.end(), [this](auto &&tag) {
+            return provided_tags_.find(tag) != provided_tags_.end();
+          });
+    };
+    for (auto it = waiting_queue_.begin(); it != waiting_queue_.end();) {
+      if (is_ready(*it)) {
+        Transaction tx = *it;
+        ReadyTransaction rtx{tx, {}};
+        updateUnlockingTransactions(rtx);
+        ready_queue_.insert({rtx.hash, rtx});
+        auto old_it = it;
+        it++;
+        waiting_queue_.erase(old_it);
+      } else {
+        it++;
+      }
+    }
+  }
+
+  void TransactionPoolImpl::updateUnlockingTransactions(
+      const ReadyTransaction &rtx) {
+    for (auto &tag : rtx.requires) {
+      if (auto it = provided_tags_.find(tag); it != provided_tags_.end()) {
+        if (auto ready_it = ready_queue_.find(it->second);
+            ready_it != ready_queue_.end()) {
+          auto &unlocking_transaction = ready_it->second;
+          unlocking_transaction.unlocks.push_back(rtx.hash);
+        }
+      }
+    }
+  }
+
+  std::vector<primitives::Transaction> TransactionPoolImpl::enforceLimits() {
+    auto [ready_num, waiting_num] = getStatus();
+    int64_t dr = ready_num - limits_.max_ready_num;
+    int64_t dw = waiting_num - limits_.max_waiting_num;
+    std::vector<Transaction> removed;
+    removed.reserve((dr > 0 ? dr : 0) + (dw > 0 ? dw : 0));
+    while (ready_num > limits_.max_ready_num) {
+      removed.push_back(ready_queue_.begin()->second);
+      ready_queue_.erase(ready_queue_.begin());
+      ready_num--;
+    }
+    while (waiting_num > limits_.max_waiting_num) {
+      removed.push_back(waiting_queue_.front());
+      waiting_queue_.pop_front();
+      waiting_num--;
+    }
+    return removed;
+  }
+
+  TransactionPoolImpl::Status TransactionPoolImpl::getStatus() const {
+    return Status{ready_queue_.size(), waiting_queue_.size()};
+  }
+
+  std::vector<Transaction> TransactionPoolImpl::pruneTags(
+      const primitives::BlockId &at, const primitives::TransactionTag &tag,
+      const std::vector<common::Hash256> &known_imported_hashes) {
+    std::list<primitives::TransactionTag> to_remove{tag};
+    std::vector<Transaction> removed;
+
+    while (not to_remove.empty()) {
+      auto tag = to_remove.front();
+      to_remove.pop_front();
+      auto hash = provided_tags_.at(tag);
+      provided_tags_.erase(tag);
+      auto tx = ready_queue_.at(hash);
+      ready_queue_.erase(hash);
+
+      auto find_previous = [this, &tx](primitives::TransactionTag const &tag)
```

Comment:
this does not look right. on every `pruneTags` you create same lambda. 
Instead, create member method (private?) and pass tx by ref there.

Reply:
Actually, I agree, don't know why it's done this way in Rust. Gonna fix it.